graphql_compiler/compiler/workarounds/orientdb_query_execution.py

```diff
@@ -0,0 +1,341 @@
+    - Must not be "optional: true".
+    - Must not have a "while:" clause nor follow a location that has one.
+    - Must have a "class:" defined. This class is used for cardinality estimation, and to
+      look for available indexes that may cover any "where:" clause that may be present.
+
+The optimizations in this file improve performance by enabling execution start points according
+to the following assumptions:
+    1. Start points with "where:" clauses that reference only local fields (i.e. not tagged values
+       from other query locations) are always better than start points without a "where:".
+       This is because the filter will have to be applied one way or the other, so we might as well
+       apply it early.
+    2. If no such start points are available, we'd like to make available as many start points
+       as possible, since we'd like OrientDB to start at the start point whose class has
+       the lowest possible cardinality.
+
+The process of applying the optimizations is as follows:
+    - Exclude and ignore all query steps that are inside a fold, optional, or recursion scope,
+      or have a "where:" clause that references a non-local (i.e. tagged) field.
+    - Find all remaining query steps with "where:" clauses that reference only local fields.
+    - If any are found, we guide our actions from assumption 1 above:
+        - Ensure they have a defined "class:" -- i.e. the OrientDB scheduler will consider them
+          valid start points.
+        - Then, prune all other query steps (ones without such "where:" clauses) by removing their
+          "class:" clause, making them invalid as query start points for OrientDB's scheduler.
+    - If none are found, we guide our actions from assumption 2 above:
+        - Ensure that all query points not inside fold, optional, or recursion scope contain
+          a "class:" clause. That increases the number of available query start points,
+          so OrientDB can choose the start point of lowest cardinality.
+"""
+
+from ..blocks import CoerceType, QueryRoot, Recurse, Traverse
+from ..expressions import ContextField, ContextFieldExistence
+from ..ir_lowering_match.utils import convert_coerce_type_and_add_to_where_block
+from ..helpers import get_one_element_collection_value
+
+
+def _is_local_filter(filter_block):
+    """Return True if the Filter block references no non-local fields, and False otherwise."""
+    # We need the "result" value of this function to be mutated within the "visitor_fn".
+    # Since we support both Python 2 and Python 3, we can't use the "nonlocal" keyword here:
+    # https://www.python.org/dev/peps/pep-3104/
+    # Instead, we use a dict to store the value we need mutated, since the "visitor_fn"
+    # can mutate state in the parent scope, but not rebind variables in it without "nonlocal".
+    # TODO(predrag): Revisit this if we drop support for Python 2.
+    result = {
+        'is_local_filter': True
+    }
```

Comment:
I think this might be a bit more friendly -- I found the class approaches a bit confusing, and there are also differences between Py2 and Py3 classes... Ideally we'll just drop support for Py2 at some point and just use `nonlocal`.